JSON Serialization breaks the structure

I have noticed that in the last release, JSON serialization is broken for more complex structure.
See below the output from the pytest diff.
```
        def test_serialize_more_complex(self):
            store = self.StoreClass()
            more_complex = b"""
    {
    "foo": {
        "foo_2": {
            "foo_3": "bar",
            "foo_4": "bar2"
        },
        "fizz" : "buzz"
        }
    }
    """
            store.parse(more_complex)
            out = BytesIO()
            store.serialize(out)

>           assert out.getvalue() == more_complex
E           assert b'{\n    "foo...: "buzz"\n}\n' == b'\n{\n"foo": ..."\n    }\n}\n'
E             At index 0 diff: 123 != 10
E             Right contains more items, first extra item: 105
E             Full diff:
E             - (b'{\n    "foo.foo_2.foo_3": "bar",\n    "foo.foo_2.foo_4": "bar2",\n    "foo.'
E             -  b'fizz": "buzz"\n}\n')
E             + (b'\n{\n"foo": {\n    "foo_2": {\n        "foo_3": "bar",\n        "foo_4": '
E             +  b'"bar2"\n    },\n    "fizz" : "buzz"\n    }\n}\n')
```

I have added this test to the pull request here
https://github.com/translate/translate/pull/3856

Basically, the json structure with several nested dictionaries is flattened to the single key "foo.foo_2.foo_3"

I would like to have this fixed, so if you any of you could point me where to start, It would be great.

Thanks, 
Jakub

--- a/translate/storage/base.py
+++ b/translate/storage/base.py
@@ -18,6 +18,7 @@
 
 """Base classes for storage interfaces."""
 
+from collections import OrderedDict
 import codecs
 import logging
 import pickle
@@ -889,38 +890,100 @@ class TranslationStore:
         return "id"
 
 
+class UnitId:
+    KEY_SEPARATOR = "."
+    INDEX_SEPARATOR = ""
+
+    def __init__(self, parts):
+        self.parts = parts
+
+    def __str__(self):
+        def fmt(element, key):
+            if element == 'key':
+                return '{}{}'.format(self.KEY_SEPARATOR, key)
+            elif element == 'index':
+                return '{}[{}]'.format(self.INDEX_SEPARATOR, key)
+            else:
+                raise ValueError('Unsupported element: {}'.format(element))
+        return ''.join([fmt(*part) for part in self.parts])
+
+    def __add__(self, other):
+        if not isinstance(other, list):
+            raise ValueError('Not supported type for add: {}'.format(type(other)))
+        return self.__class__(self.parts + other)
+
+    @classmethod
+    def from_string(cls, text):
+        result = []
+        # Strip possible leading separator
+        if text.startswith(cls.KEY_SEPARATOR):
+            text = text[len(cls.KEY_SEPARATOR):]
+        for item in text.split(cls.KEY_SEPARATOR):
+            if '[' in item and item[-1] == ']':
+                item, pos = item[:-1].split('[')
+                if cls.INDEX_SEPARATOR and item:
+                    result.append(("key", item))
+                result.append(("index", int(pos)))
+            else:
+                result.append(("key", item))
+        return cls(result)
+
+
 class DictUnit(TranslationUnit):
+    IdClass = UnitId
+    DefaultDict = OrderedDict
+
+    def __init__(self, source=None):
+        super().__init__(source)
+        self._unitid = None
+
+    def storevalue(self, output, value, override_key=None):
+        target = output
+        if self._unitid is None:
+            self._unitid = self.IdClass.from_string(self._id)
+        parts = self._unitid.parts
+        for pos, part in enumerate(parts[:-1]):
+            element, key = part
+            default = [] if parts[pos + 1][0] == 'index' else self.DefaultDict()
+            if element == 'index':
+                if len(target) <= key:
+                    target.append(default)
+            elif element == 'key':
+                if key not in target or isinstance(target[key], str):
+                    target[key] = default
+            else:
+                raise ValueError('Unsupported element: {}'.format(element))
+            target = target[key]
+        if override_key:
+            element, key = 'key', override_key
+        else:
+            element, key = parts[-1]
+        if element == 'key':
+            target[key] = value
+        elif element == 'index':
+            if len(target) <= key:
+                target.append(value)
+            else:
+                target[key] = value
+        else:
+            raise ValueError('Unsupported element: {}'.format(element))
+
+    def storevalues(self, output):
+        self.storevalue(output, self.value)
+
     def getvalue(self):
         """Returns dictionary for serialization."""
-        raise NotImplementedError()
+        result = {}
+        self.storevalues(result)
+        return result
+
+    def set_unitid(self, unitid):
+        self.setid(str(unitid))
+        self._unitid = unitid
 
 
 class DictStore(TranslationStore):
-    def serialize_merge(self, d1, d2):
-        for k in d2:
-            if k in d1:
-                if isinstance(d1[k], dict) and isinstance(d2[k], dict):
-                    self.serialize_merge(d1[k], d2[k])
-                elif isinstance(d1[k], list) and isinstance(d2[k], tuple):
-                    if isinstance(d2[k][1], dict):
-                        if len(d1[k]) > d2[k][0]:
-                            d1[k][d2[k][0]].update(d2[k][1])
-                        else:
-                            d1[k].append(d2[k][1])
-                    else:
-                        if len(d1[k]) > d2[k][0]:
-                            d1[k][d2[k][0]] = d2[k][1]
-                        else:
-                            d1[k].append(d2[k][1])
-                elif isinstance(d1[k], list) and isinstance(d2[k], list):
-                    d1[k].extend(d2[k])
-                else:
-                    d1[k] = d2[k]
-            elif isinstance(d2[k], tuple):
-                d1[k] = [d2[k][1]]
-            else:
-                d1[k] = d2[k]
 
     def serialize_units(self, output):
         for unit in self.unit_iter():
-            self.serialize_merge(output, unit.getvalue())
+            unit.storevalues(output)

--- a/translate/storage/jsonl10n.py
+++ b/translate/storage/jsonl10n.py
@@ -119,18 +119,14 @@ class JsonUnit(base.DictUnit):
         """Converts to a string representation."""
         return json.dumps(self.getvalue(), separators=(',', ': '), indent=4, ensure_ascii=False)
 
-    def getkey(self):
-        return self.getid().lstrip('.')
-
     def converttarget(self):
         if issubclass(self._type, str):
             return self.target
         else:
             return self._type(self.target)
 
-    def getvalue(self):
-        """Return value to be stored in JSON file."""
-        return {self.getkey(): self.converttarget()}
+    def storevalues(self, output):
+        self.storevalue(output, self.converttarget())
 
 
 class JsonFile(base.DictStore):
@@ -158,7 +154,7 @@ class JsonFile(base.DictStore):
         out.write(json.dumps(units, **self.dump_args).encode(self.encoding))
         out.write(b'\n')
 
-    def _extract_units(self, data, stop=None, prev="", name_node=None, name_last_node=None, last_node=None):
+    def _extract_units(self, data, stop=None, prev=None, name_node=None, name_last_node=None, last_node=None):
         """Recursive function to extract items from the data files
 
         :param data: the current branch to walk down
@@ -168,13 +164,15 @@ class JsonFile(base.DictStore):
         :param name_last_node: the name of the last node
         :param last_node: the last list or dict
         """
+        if prev is None:
+            prev = self.UnitClass.IdClass([])
         if isinstance(data, dict):
             for k, v in data.items():
-                for x in self._extract_units(v, stop, "%s.%s" % (prev, k), k, None, data):
+                for x in self._extract_units(v, stop, prev + [('key', k)], k, None, data):
                     yield x
         elif isinstance(data, list):
             for i, item in enumerate(data):
-                for x in self._extract_units(item, stop, "%s[%s]" % (prev, i), i, name_node, data):
+                for x in self._extract_units(item, stop, prev + [('index', i)], i, name_node, data):
                     yield x
         # apply filter
         elif (stop is None or
@@ -182,7 +180,7 @@ class JsonFile(base.DictStore):
               (isinstance(last_node, list) and name_last_node in stop)):
 
             unit = self.UnitClass(data, name_node)
-            unit.setid(prev)
+            unit.set_unitid(prev)
             yield unit
 
     def parse(self, input):
@@ -211,17 +209,8 @@ class JsonFile(base.DictStore):
 
 class JsonNestedUnit(JsonUnit):
 
-    def getkey(self):
-        return self.getid().lstrip('.').split('.')
-
-    def getvalue(self):
-        ret = self.converttarget()
-        for k in reversed(self.getkey()):
-            if '[' in k and k[-1] == ']':
-                k, pos = k[:-1].split('[')
-                ret = (int(pos), ret)
-            ret = OrderedDict({k: ret})
-        return ret
+    def storevalues(self, output):
+        self.storevalue(output, self.converttarget())
 
 
 class JsonNestedFile(JsonFile):
@@ -231,7 +220,7 @@ class JsonNestedFile(JsonFile):
 
 
 class WebExtensionJsonUnit(JsonUnit):
-    def getvalue(self):
+    def storevalues(self, output):
         value = OrderedDict((
             ('message', self.target),
         ))
@@ -239,7 +228,7 @@ class WebExtensionJsonUnit(JsonUnit):
             value['description'] = self.notes
         if self.placeholders:
             value['placeholders'] = self.placeholders
-        return {self.getid(): value}
+        self.storevalue(output, value)
 
 
 class WebExtensionJsonFile(JsonFile):
@@ -253,7 +242,7 @@ class WebExtensionJsonFile(JsonFile):
 
     UnitClass = WebExtensionJsonUnit
 
-    def _extract_units(self, data, stop=None, prev="", name_node=None, name_last_node=None, last_node=None):
+    def _extract_units(self, data, stop=None, prev=None, name_node=None, name_last_node=None, last_node=None):
         for item, value in data.items():
             unit = self.UnitClass(
                 value.get('message', ''),
@@ -303,18 +292,12 @@ class I18NextUnit(JsonNestedUnit):
         self._rich_target = None
         self._target = target
 
-    def getvalue(self):
+    def storevalues(self, output):
         if not isinstance(self.target, multistring):
-            return super().getvalue()
-
-        ret = OrderedDict()
-        for i, value in enumerate(self.target.strings):
-            ret[self._item[i]] = value
-
-        path = self.getid().lstrip('.').split('.')[:-1]
-        for k in reversed(path):
-            ret = {k: ret}
-        return ret
+            super().storevalues(output)
+        else:
+            for i, value in enumerate(self.target.strings):
+                self.storevalue(output, value, override_key=self._item[i])
 
 
 class I18NextFile(JsonNestedFile):
@@ -325,7 +308,9 @@ class I18NextFile(JsonNestedFile):
 
     UnitClass = I18NextUnit
 
-    def _extract_units(self, data, stop=None, prev="", name_node=None, name_last_node=None, last_node=None):
+    def _extract_units(self, data, stop=None, prev=None, name_node=None, name_last_node=None, last_node=None):
+        if prev is None:
+            prev = self.UnitClass.IdClass([])
         if isinstance(data, dict):
             plurals_multiple = [key.rsplit('_', 1)[0] for key in data if key.endswith('_0')]
             plurals_simple = [key.rsplit('_', 1)[0] for key in data if key.endswith('_plural')]
@@ -359,11 +344,12 @@ class I18NextFile(JsonNestedFile):
                         sources.append(data[key])
                         items.append(key)
                     unit = self.UnitClass(multistring(sources), items)
-                    unit.setid("%s.%s" % (prev, plural_base))
+                    newid = prev + [('key', plural_base)]
+                    unit.set_unitid(newid)
                     yield unit
                     continue
 
-                for x in self._extract_units(v, stop, "%s.%s" % (prev, k), k, None, data):
+                for x in self._extract_units(v, stop, prev + [('key', k)], k, None, data):
                     yield x
         else:
             parent = super()._extract_units(
@@ -414,7 +400,7 @@ class GoI18NJsonFile(JsonFile):
             locale = "en"
         return plural_tags.get(locale, plural_tags['en'])
 
-    def _extract_units(self, data, stop=None, prev="", name_node=None, name_last_node=None, last_node=None):
+    def _extract_units(self, data, stop=None, prev=None, name_node=None, name_last_node=None, last_node=None):
         for value in data:
             translation = value.get('translation', '')
             if isinstance(translation, dict):
@@ -443,16 +429,16 @@ class ARBJsonUnit(JsonUnit):
         super().__init__(source, item, notes, placeholders, **kwargs)
         self.metadata = metadata or {}
 
-    def getvalue(self):
+    def storevalues(self, output):
         if self.notes:
             self.metadata['description'] = self.notes
         identifier = self.getid()
         if identifier == "@":
-            return self.metadata
-        return OrderedDict((
-            (identifier, self.target),
-            ('@{}'.format(identifier), self.metadata),
-        ))
+            for key, value in self.metadata.items():
+                self.storevalue(output, value, override_key=key)
+        else:
+            self.storevalue(output, self.target, override_key=identifier)
+            self.storevalue(output, self.metadata, override_key='@{}'.format(identifier))
 
     def isheader(self):
         return self._id == "@"
@@ -477,7 +463,7 @@ class ARBJsonFile(JsonFile):
             'ensure_ascii': False,
         }
 
-    def _extract_units(self, data, stop=None, prev="", name_node=None, name_last_node=None, last_node=None):
+    def _extract_units(self, data, stop=None, prev=None, name_node=None, name_last_node=None, last_node=None):
         # Extract metadata as header
         metadata = OrderedDict([(key, value) for key, value in data.items() if key.startswith("@@")])
         if metadata:

--- a/translate/storage/test_jsonl10n.py
+++ b/translate/storage/test_jsonl10n.py
@@ -52,6 +52,25 @@ JSON_ARRAY = b"""{
     ]
 }
 """
+JSON_COMPLEX = b"""{
+    "key": "value",
+    "key.key": "value",
+    "key[0]": "value2",
+    "key3": [
+        "one",
+        "two"
+    ],
+    "key4": [
+        {
+            "nested": "one"
+        },
+        [
+            "one",
+            "two"
+        ]
+    ]
+}
+"""
 JSON_GOI18N = b"""[
     {
         "id": "tag",
@@ -154,6 +173,15 @@ class TestJSONResourceStore(test_monolingual.TestMonolingualStore):
         out = BytesIO()
         store.serialize(out)
         assert out.getvalue() == content
+
+    def test_complex(self):
+        store = self.StoreClass()
+        store.parse(JSON_COMPLEX)
+
+        out = BytesIO()
+        store.serialize(out)
+
+        assert out.getvalue() == JSON_COMPLEX
 
 
 class TestJSONNestedResourceStore(test_monolingual.TestMonolingualUnit):

--- a/translate/storage/test_yaml.py
+++ b/translate/storage/test_yaml.py
@@ -417,3 +417,20 @@ class TestRubyYAMLResourceStore(test_monolingual.TestMonolingualStore):
         store.parse(data)
         assert len(store.units) == 5
         assert bytes(store).decode('ascii') == data
+
+    def test_type_change(self):
+        original = '''en:
+  days_on: '["Sunday", "Monday"]'
+'''
+        changed = '''en:
+  days_on:
+  - Sunday
+  - Monday
+'''
+        store = self.StoreClass()
+        store.parse(original)
+        update = self.StoreClass()
+        update.parse(changed)
+        for unit in update.units:
+            store.addunit(unit)
+        assert bytes(store).decode('ascii') == changed

--- a/translate/storage/yaml.py
+++ b/translate/storage/yaml.py
@@ -29,11 +29,27 @@ from translate.misc.multistring import multistring
 from translate.storage import base
 
 
+class YAMLUnitId(base.UnitId):
+    KEY_SEPARATOR = "->"
+    INDEX_SEPARATOR = "->"
+
+    def __str__(self):
+        result = super().__str__()
+        # Strip leading ->
+        if result.startswith(self.KEY_SEPARATOR):
+            return result[len(self.KEY_SEPARATOR):]
+        return result
+
+
 class YAMLUnit(base.DictUnit):
     """A YAML entry"""
 
+    IdClass = YAMLUnitId
+    DefaultDict = dict
+
     def __init__(self, source=None, **kwargs):
-        self._id = None
+        # Ensure we have ID (for serialization)
+        self._id = str(uuid.uuid4())
         if source:
             self.source = source
         super().__init__(source)
@@ -50,30 +66,16 @@ class YAMLUnit(base.DictUnit):
         self._id = value
 
     def getid(self):
-        # Ensure we have ID (for serialization)
-        if self._id is None:
-            self._id = str(uuid.uuid4())
         return self._id
 
     def getlocations(self):
         return [self.getid()]
 
-    def getkey(self):
-        return self.getid().split('->')
-
     def convert_target(self):
         return self.target
 
-    def getvalue(self):
-        ret = self.convert_target()
-        for k in reversed(self.getkey()):
-            if '[' in k and k[-1] == ']':
-                k, pos = k[:-1].split('[')
-                ret = (int(pos), ret)
-                if not k:
-                    continue
-            ret = {k: ret}
-        return ret
+    def storevalues(self, output):
+        self.storevalue(output, self.convert_target())
 
 
 class YAMLFile(base.DictStore):
@@ -121,12 +123,14 @@ class YAMLFile(base.DictStore):
                     'Key not string: {0}/{1} ({2})'.format(prev, k, type(k))
                 )
 
-            for x in self._flatten(v, '->'.join((prev, k)) if prev else k):
+            for x in self._flatten(v, prev + [('key', k)]):
                 yield x
 
-    def _flatten(self, data, prev=""):
+    def _flatten(self, data, prev=None):
         """Flatten YAML dictionary.
         """
+        if prev is None:
+            prev = self.UnitClass.IdClass([])
         if isinstance(data, dict):
             for x in self._parse_dict(data, prev):
                 yield x
@@ -137,8 +141,7 @@ class YAMLFile(base.DictStore):
                 yield (prev, str(data))
             elif isinstance(data, list):
                 for k, v in enumerate(data):
-                    key = '[{0}]'.format(k)
-                    for value in self._flatten(v, '->'.join((prev, key))):
+                    for value in self._flatten(v, prev + [('index', k)]):
                         yield value
             elif data is None:
                 pass
@@ -176,7 +179,7 @@ class YAMLFile(base.DictStore):
 
         for k, data in self._flatten(content):
             unit = self.UnitClass(data)
-            unit.setid(k)
+            unit.set_unitid(k)
             self.addunit(unit)
 
 

